src/plugins/es_ui_shared/static/forms/hook_form_lib/hooks/use_field.ts

```diff
@@ -0,0 +1,378 @@
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { useState, useEffect, useRef } from 'react';
+
+import {
+  Form,
+  Field,
+  FieldConfig,
+  FieldValidateResponse,
+  ValidationConfig,
+  ValidationError,
+} from '../types';
+import { FIELD_TYPES, VALIDATION_TYPES } from '../constants';
+
+export const useField = (form: Form, path: string, config: FieldConfig = {}) => {
+  const {
+    type = FIELD_TYPES.TEXT,
+    defaultValue = '',
+    label = '',
+    helpText = '',
+    validations = [],
+    formatters = [],
+    fieldsToValidateOnChange = [path],
+    isValidationAsync = false,
+    errorDisplayDelay = form.options.errorDisplayDelay,
+    serializer = (value: unknown) => value,
+    deSerializer = (value: unknown) => value,
+  } = config;
+
+  const [value, setStateValue] = useState(
+    typeof defaultValue === 'function' ? deSerializer(defaultValue()) : deSerializer(defaultValue)
+  );
+  const [errors, setErrors] = useState<ValidationError[]>([]);
+  const [isPristine, setPristine] = useState(true);
+  const [isValidating, setValidating] = useState(false);
+  const [isUpdating, setIsUpdating] = useState(false);
+  const validateCounter = useRef(0);
+  const debounceTimeout = useRef<NodeJS.Timeout | null>(null);
+
+  // -- HELPERS
+  // ----------------------------------
+
+  /**
+   * Filter an array of errors with specific validation type on them
+   *
+   * @param _errors The array of errors to filter
+   * @param validationType The validation type to filter out
+   */
+  const filterErrors = (
+    _errors: ValidationError[],
+    validationTypeToFilterOut: string | string[] = VALIDATION_TYPES.FIELD
+  ): ValidationError[] => {
+    const validationTypeToArray = Array.isArray(validationTypeToFilterOut)
+      ? (validationTypeToFilterOut as string[])
+      : ([validationTypeToFilterOut] as string[]);
+
+    return _errors.filter(error =>
+      validationTypeToArray.every(_type => error.validationType !== _type)
+    );
+  };
+
+  const runFormatters = (input: unknown): unknown => {
+    const isEmptyString = typeof input === 'string' && input.trim() === '';
+
+    if (isEmptyString) {
+      return input;
+    }
+    return formatters.reduce((output, formatter) => formatter(output), input);
+  };
+
+  const onValueChange = () => {
+    if (isPristine) {
```

Comment:
Currently, with hooks, each setState() triggers a re-render. To align both, I added a condition in use_form to avoid an unnecessary re-render once the form has been submitted.

```js
if (!isSubmitted) {
  setSubmitted(true); // User has attempted to submit the form at least once
}
```